patroni/postgresql/callback_executor.py

```diff
@@ -1,40 +1,44 @@
-import logging
-import subprocess
 from threading import Event, Lock, Thread
+import subprocess
+import logging
+import time
+import queue
 
 logger = logging.getLogger(__name__)
 
-
 class CallbackExecutor(Thread):
+    '''
+    Execution thread that queues commands and executes them in order.
+    Guarantees that commands are executed in the order they are received via
+    FIFO queue.
+    '''
 
-    def __init__(self):
+    def __init__(self, executor=subprocess.Popen):
         super(CallbackExecutor, self).__init__()
         self.daemon = True
-        self._lock = Lock()
-        self._cmd = None
-        self._process = None
-        self._callback_event = Event()
+        self._cmds_queue = queue.Queue()
+        self._executor = executor
         self.start()
 
     def call(self, cmd):
-        with self._lock:
-            if self._process and self._process.poll() is None:
-                try:
-                    self._process.kill()
-                    logger.warning('Killed the old callback process because it was still running: %s', self._cmd)
-                except OSError:
-                    logger.exception('Failed to kill the old callback')
-        self._cmd = cmd
-        self._callback_event.set()
+      # Queue up the command
+      self._cmds_queue.put(cmd)
 
     def run(self):
         while True:
-            self._callback_event.wait()
-            self._callback_event.clear()
-            with self._lock:
-                try:
-                    self._process = subprocess.Popen(self._cmd, close_fds=True)
-                except Exception:
-                    logger.exception('Failed to execute %s',  self._cmd)
-                    continue
-            self._process.wait()
+            logger.info("Callback sleeping until new command is queued")
+            cmd = self._cmds_queue.get()
+            cmd_str = ' '.join(cmd)
+            logger.info("Callback executor waking to call {}".format(cmd_str))
+            try:
+                logger.info("Spawning process for {}".format(cmd_str))
+                p = self._executor(cmd, close_fds=True)
```

Comment:
This made it much easier to test. I couldn't get the `@patch` decorator to work.

Reply:
I think `pip install coverage flake8 mock pytest-cov pytest` should help.
After that you have to:
1. Run `python setup.py test` -- to execute all unit tests
2. Run `python setup.py flake8` -- to check code formatting

Reply:
I'm able to run the tests, but the patch was not actually patching. This works and is much more straight forward to test.
Any reason not to leave it this way?

Reply:
> but the patch was not actually patching

And such behavior has a reasonable explanation.
It happens due to the executor being instantiated [here](https://github.com/zalando/patroni/pull/1239/files#diff-452cfdd642bafda8a084b3557b9dcc9aR15), during the [import](https://github.com/zalando/patroni/pull/1239/files#diff-73d9a9b641bebccdf53fe41ba6630a94R4), which happens earlier than the test is being executed.

> Any reason not to leave it this way?

Previously there was a good reason to use `subprocess.Popen`, because we wanted to [kill](https://github.com/zalando/patroni/pull/1239/files#diff-452cfdd642bafda8a084b3557b9dcc9aL23) the old callback, now I would really like just use `subprocess.call`.

Actually the whole topic of callbacks is way more complicated than one could think about it due to the nature of distributed systems: https://github.com/zalando/patroni/issues/536
If you just update labels on the pod with a callback it might be fine, but as long as you start updating the endpoint, serialization will make everything worse and you may easily end up on having the endpoint pointing to the wrong pod.

By the way, the existing implementation of callback executor actually _guarantees_ that only one callback script is running at a time.  Another propery of it that it automatically skips intermediate calls while there is one script running.

Can you actually share the code of your callback script and patroni config file?

Reply:
The existing implementation guaranteed that you couldn't run 2x callbacks at the same time, but there is a bug since `self._cmd` is set outside of the lock. This means that prior to subprocess.Popen being called, the command could be set to a different command if another callback were to be executed.

Here is the callback we are using now
https://github.com/zalando/spilo/blob/1.5-p5/postgres-appliance/scripts/callback_role.py

Reply:
> there is a bug since self._cmd is set outside of the lock. This means that prior to subprocess.Popen being called, the command could be set to a different command if another callback were to be executed.

Nope, it is absolutely fine to set it without a lock. Acess is sort of protected by `Event()` object.
After all, we are interested in the LAST callback.

> Here is the callback we are using now

Now I see what the problem is. There is one case when callback_role.py is executed not directly by Patroni, but from the shell script https://github.com/zalando/spilo/blob/master/postgres-appliance/scripts/configure_spilo.py#L275.
When Patroni is [killing](https://github.com/zalando/patroni/pull/1239/files#diff-452cfdd642bafda8a084b3557b9dcc9aL23) the currently running script, it just terminates the parent process, but all subprocesses continue to run. After that it executes the callback script one more time. BOOM! We have got two instances of callback_role.py running.

There is a good reason why we choose to terminate the old instance of the callback. Depending on how it (callback script) implements retry, it might be quite a long-running process. While it is running the Patroni state already changed and it doesn't make much sence to continue, because it just delays the correct callback.

The good solution I could think about is:
1. In the `call()` method collect all offsprings of the callback process
2. kill the callback process (it is already there)
3. kill offsprings
4. In the `run()` method wait for the callback process to finish (it is already there)
5. After that wait for offsprings to finish (if there are any)